kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/KafkaProxyInitializer.java

```diff
@@ -173,8 +187,41 @@ private void addHandlers(SocketChannel ch, VirtualClusterBinding binding) {
         }
 
         ApiVersionsServiceImpl apiVersionService = new ApiVersionsServiceImpl();
-        var frontendHandler = new KafkaProxyFrontendHandler(context -> {
-            List<FilterAndInvoker> apiVersionFilters = dp.isAuthenticationOffloadEnabled() ? List.of()
+        final NetFilter netFilter = new InitalizerNetFilter(dp, apiVersionService, ch, virtualCluster, binding, pfr, filterChainFactory, endpointReconciler);
+        var frontendHandler = new KafkaProxyFrontendHandler(netFilter, dp, virtualCluster, apiVersionService);
+
+        pipeline.addLast("netHandler", frontendHandler);
+
+        LOGGER.debug("{}: Initial pipeline: {}", ch, pipeline);
+    }
+
+    @VisibleForTesting
+    static class InitalizerNetFilter implements NetFilter {
+
+        private final SaslDecodePredicate decodePredicate;
+        private final ApiVersionsServiceImpl apiVersionService;
+        private final SocketChannel ch;
+        private final VirtualCluster virtualCluster;
+        private final VirtualClusterBinding binding;
+        private final PluginFactoryRegistry pfr;
+        private final FilterChainFactory filterChainFactory;
+        private final EndpointReconciler endpointReconciler;
+
+        InitalizerNetFilter(SaslDecodePredicate decodePredicate, ApiVersionsServiceImpl apiVersionService, SocketChannel ch, VirtualCluster virtualCluster,
+                            VirtualClusterBinding binding, PluginFactoryRegistry pfr, FilterChainFactory filterChainFactory, EndpointReconciler endpointReconciler) {
+            this.decodePredicate = decodePredicate;
+            this.apiVersionService = apiVersionService;
+            this.ch = ch;
+            this.virtualCluster = virtualCluster;
```

Comment:
superficial improvement
```suggestion
            this.virtualCluster = binding.virtualCluster();
```

Reply:
Yes but its enough to keep sonar happy... 😩